### DISCUSSION TITLE
Release assets don't have persistent timestamps

### DISCUSSION THREAD

**Discussion by gasinvein** [community/community#8546] [Other Feature Feedback, Questions, & Ideas]
I've notices that release assets, both auto-generated and uploaded, now have HTTP `Date` header set to current date/time.
Is this intentional?
Such behaviour breaks workflows that rely on the downloaded asset timestamp to determine if it should be re-downloaded, e.g. when using `wget` in makefiles.

**Comment by MylesBorins** [comment]
I'm going to bring this up to engineering to see what is going on here.

**Reply by WillHalto** [reply]
👋 @gasinvein I am taking a look here, just to confirm, are you referring to the `date` header or the `last-modified` header? And could you elaborate briefly on why it has this impact:

> Such behaviour breaks workflows that rely on the downloaded asset timestamp to determine if it should be re-downloaded, e.g. when using wget in makefiles.

Thanks!

**Reply by gasinvein** [reply]
I'm talking about `Date`. Looks like the date in `Last-Modified` header isn't changing, but `Last-Modified` is set only for uploaded assets, not for automatically generated on GitHub ones.
>why it has this impact

Some downloader tools, like `wget`, use the HTTP headers to set mtime on downloaded files. If such download is used in a Makefile as a dependency, and HTTP is always set to the current date, the downloaded file becomes newer than anything that depends on it, forcing otherwise unnecessary rebuilds.

**Reply by WillHalto** [reply]
@gasinvein gotcha, thank you for that info! Unless I am mistaken, I believe that's the correct behavior of the HTTP `date` header - it's a timestamp of when the response was sent. 

As for why the Github generated tarballs don't have a `last-modified` date, that's because they are generated on demand when the download is requested.

**Reply by gasinvein** [reply]
> As for why the Github generated tarballs don't have a `last-modified` date, that's because they are generated on demand when the download is requested.

Aren't they cached? Even if not, it feels like it makes sense for them to have `Last-Modified` timestamp, since they aren't expected to change over time after being generated for the first time.

**Reply by fenollp** [reply]
I also ask for `Last-Modified` header to be set on github-generated tarballs.
Reproducibly generated tarballs or archives should help in this endeavor and in fact I argue this is the underlying issue being discussed.
This would help for content-addressed caches such as Bazel's and of other build systems.

Here's [my take on making these generated tarballs hermetic](https://github.com/fenollp/deterministic-archive-proxy). Basically, it boils down to first sorting the paths that will be tar-ed and compressed.
